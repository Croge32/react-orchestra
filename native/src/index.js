// require("babel-core/register");
// require("babel-polyfill");

import Note from './components/Note';
import Instrument from './components/Instrument';
import Orchestra from './components/Orchestra';

export default {
  Instrument,
  Note,
  Orchestra,
};

export {
  Instrument,
  Note,
  Orchestra,
};
