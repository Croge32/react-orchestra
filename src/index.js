import React from 'react'

import {
  Note,
  Instrument,
  Orchestra,
} from './platforms/web/';

export default {
  Instrument,
  Note,
  Orchestra,
}

export {
  Instrument,
  Note,
  Orchestra,
};


// class App extends React.Component {
//   render() {
//     return <div>
//       <Note />
//       <Instrument />
//       <Orchestra />
//       Orchestra web main
//     </div>
//   }
// }
// export default App;
