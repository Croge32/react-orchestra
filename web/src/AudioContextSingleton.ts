const getAudioContext = () => {
  const AudioContext =
    // @ts-ignore
    window.AudioContext || // Default
    // @ts-ignore
    window.webkitAudioContext || // Safari and old versions of Chrome
    false;
  if (!AudioContext) {
    console.warn(
      "Sorry but the WebAudio API is not supported on this browser. Please consider using Chrome or Safari for the best experience "
    );
    return {};
    // throw new Error('PLATFORM_NOT_SUPPORTED');
  }
  return new AudioContext();
};

const audioContext = getAudioContext();
export default audioContext;
