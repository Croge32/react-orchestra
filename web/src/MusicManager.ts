// import { noteNamesWithOctaves } from '../../constants/NOTE_NAMES';
// import Sound from './Sound';
import audioContext from "./AudioContextSingleton";
import MidiIO from "./midi-io/";
import Store from "./stores";
import { Midi } from "./types";

const baseURL =
  "https://raw.githubusercontent.com/RakanNimer/midi-js-soundfonts/master/MusyngKite";
// const playingNotes = {};

export type InstrumentName = string;

export type NoteName = string;

const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
const instrumentNameToRemotePath = (instrumentName: InstrumentName) =>
  `${baseURL}/${instrumentName}-mp3`;
const instrumentAndNoteNameToRemotePath = (
  instrumentName: InstrumentName,
  noteName: NoteName
) => `${baseURL}/${instrumentName}-mp3/${noteName}.mp3`;

const generateNoteKey = (instrumentName: InstrumentName, noteName: NoteName) =>
  `${instrumentName}-${noteName}`;

const getSoundFromRemote = async (
  instrumentName: InstrumentName,
  noteName: NoteName
) => {
  const noteURL = instrumentAndNoteNameToRemotePath(instrumentName, noteName);
  const mp3 = await fetch(noteURL);
  const mp3Blob = await mp3.arrayBuffer();
  return mp3Blob;
};

const getNoteBlob = async (
  instrumentName: InstrumentName,
  noteName: NoteName
) => {
  const noteKey = generateNoteKey(instrumentName, noteName);
  const isNoteCached = await Store.exists(noteKey);
  if (isNoteCached) {
    const item = await Store.get(noteKey);
    return item;
  }
  const noteBlob = await getSoundFromRemote(instrumentName, noteName);
  return noteBlob;
};

const loadSound = async (
  instrumentName: InstrumentName,
  noteName: NoteName
) => {
  const noteBlob = await getNoteBlob(instrumentName, noteName);
  return noteBlob;
};

const playSound = (noteBlob: any, { gain = 1 }) =>
  new Promise((resolve, reject) => {
    audioContext.decodeAudioData(
      noteBlob,
      async (buffer: any) => {
        // create a sound source
        const source = audioContext.createBufferSource();
        // tell the source which sound to play
        source.buffer = buffer;
        // connect the source to the context's destination (the speakers)
        source.connect(audioContext.destination);
        // Create a gain node.
        const gainNode = audioContext.createGain();
        source.connect(gainNode);
        gainNode.connect(audioContext.destination);
        // Reduce the volume.
        gainNode.gain.value = gain;
        source.start(0);
        resolve(source);
      },
      (err: any) => {
        reject(err);
      }
    );
  });

const playNote = async (
  instrumentName: InstrumentName,
  noteName: NoteName,
  { gain = 1 }
) => {
  const noteBlob = await getNoteBlob(instrumentName, noteName);
  return await playSound(noteBlob, { gain });
};

const stopPlayingNote = async (noteBuffer: any, fadeOutDuration = 700) => {
  try {
    await delay(fadeOutDuration);
    noteBuffer.stop();
  } catch (err) {
    console.warn(`ERROR stopping note playback ${err.message}`);
  }
};

const sharpToBemol = (noteName: NoteName) => {
  if (noteName.indexOf("#") > -1) {
    const noteNameNoOctave = noteName[0];
    const octave = noteName[noteName.length - 1];
    const noteCharCode = noteNameNoOctave.charCodeAt(0);
    const newNote =
      noteNameNoOctave !== "G"
        ? `${String.fromCharCode(noteCharCode + 1)}b${octave}`
        : `Ab${octave}`;
    return newNote;
  }
  return noteName;
};

const midiURLToMetaAndTracks = async (midiURL: string) => {
  // TODO : cache
  const parsedMidi = await MidiIO.parseMidi(midiURL);
  const { meta, tracks } = MidiIO.getTracksAndMetaFromParsedMidi(parsedMidi);
  return { meta, tracks };
};

export {
  generateNoteKey,
  midiURLToMetaAndTracks,
  playSound,
  playNote,
  sharpToBemol,
  instrumentNameToRemotePath,
  stopPlayingNote,
  getNoteBlob,
  getSoundFromRemote,
  loadSound,
  delay
};

export default {
  generateNoteKey,
  midiURLToMetaAndTracks,
  playSound,
  playNote,
  sharpToBemol,
  instrumentNameToRemotePath,
  stopPlayingNote,
  getNoteBlob,
  getSoundFromRemote,
  loadSound,
  delay
};
