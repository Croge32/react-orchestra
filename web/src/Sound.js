class Sound {
  constructor(localPath) {
    this.id = localPath;
    this.localPath = localPath;
    return this;
  }
  async load(remotePath) {
    const note = await fetch(remotePath);
    return { note, localPath: this.localPath };
  }
}
export default Sound;
