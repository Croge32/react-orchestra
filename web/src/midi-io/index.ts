import defaultsDeep from "lodash/defaultsDeep";
import * as MidiUtils from "./midi-utils";

type Midi = any;

export default class MidiIO {
  millisecondsPerTick?: number;
  timeSignatureNumerator?: number;
  timeSignatureDenominator?: number;
  ticksPerBeat?: number;
  BPM?: number;
  settings = {
    midiURL: null as string | null,
    parsedMidi: null as Midi | null,
    BPM: -1 as number
  };
  isInitialised = false;

  constructor(userSettings: MidiIO["settings"]) {
    this.settings = defaultsDeep(userSettings, {
      midiURL: null,
      parsedMidi: null,
      BPM: -1
    });
  }
  setBPM(BPM = 60) {
    this.millisecondsPerTick = MidiIO.getMillisecondsPerTick(
      MidiIO.bpmToMSPerBeat(
        BPM,
        this.timeSignatureNumerator,
        this.timeSignatureDenominator
      ),
      // @ts-ignore
      this.ticksPerBeat
    );
    this.BPM = BPM;
  }

  static parseMidi = MidiUtils.parseMidi;
  static getTracksAndMetaFromParsedMidi =
    MidiUtils.getTracksAndMetaFromParsedMidi;
  static getWriter = MidiUtils.getWriter;
  static getUniqueFromMidiNotes = MidiUtils.getUniqueFromMidiNotes;
  static midiNoteNumberToName = MidiUtils.midiNoteNumberToName;
  static urlToJSON = MidiUtils.urlToJSON;
  static getAllMetadata = MidiUtils.getAllMetadata;
  static getAllTracks = MidiUtils.getAllTracks;
  static getInstrumentNumberFromMidiTrack =
    MidiUtils.getInstrumentNameFromMidiTrack;
  static binaryStringToJSON = MidiUtils.binaryStringToJSON;
  static getNoteOnEvents = MidiUtils.getNoteOnEvents;
  static getNoteOffEvents = MidiUtils.getNoteOffEvents;
  static getTempoFromTrack = MidiUtils.getTempoFromTrack;
  static getMetadataFromTrack = MidiUtils.getMetadataFromTrack;
  static urlToBinaryString = MidiUtils.urlToBinaryString;
  static noteOffEventToNote = MidiUtils.noteOffEventToNote;
  static getMillisecondsPerTick = MidiUtils.getMillisecondsPerTick;
  static getBPMData = MidiUtils.getBPMData;
  static bpmToMSPerBeat = MidiUtils.bpmToMSPerBeat;
}
