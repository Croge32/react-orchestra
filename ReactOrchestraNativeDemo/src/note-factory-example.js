import React from 'react';
import { View } from 'react-native';
import { NoteFactory } from 'react-orchestra/native';

const renderNote = (instrumentName, noteName) => <View style={{ backgroundColor: 'papayawhip' }}> I am a note : {instrumentName} {noteName} You can click me ! </View>;

class NoteFactoryExample extends React.Component {
  render() {
    console.warn(JSON.stringify(Object.keys(NoteFactory), 2, 2));
    return (
      <NoteFactory
        type="scale"
        scaleName="ionian"
        noteName="C"
        instrumentName="acoustic_grand_piano"
        startOctave="3"
        octaveCount="1"
        renderNote={renderNote}
      />
    );
  }
}
export default NoteFactoryExample;
