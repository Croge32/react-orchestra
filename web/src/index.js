import Instrument from './components/Instrument';
import Note from './components/Note';
import Orchestra from './components/Orchestra';
import NoteFactory from './components/NoteFactory';
import KeyBinding from './components/KeyBinding';
// import NoteHelpers from './utils/NoteHelpers';
import { InstrumentHelpers } from './InstrumentJS/src/';
import MusicManager from './MusicManager';

window.isTouchDevice = 'ontouchstart' in document.documentElement;

export default {
  Instrument,
  Note,
  Orchestra,
  NoteFactory,
  KeyBinding,
  InstrumentHelpers,
  MusicManager,
//  NoteHelpers
};

export {
  Instrument,
  Note,
  Orchestra,
  NoteFactory,
  KeyBinding,
  InstrumentHelpers,
  MusicManager,
//  NoteHelpers
};
