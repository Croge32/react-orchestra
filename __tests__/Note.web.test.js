
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';

import { Note } from '../src/';

const note = (
  <Note
    name={'A3'}
    play={false}
    instrumentName={'acoustic_grand_piano'}
    onNoteLoaded={()=>{
      console.log('Note loaded');
    }}
  >
    <div>Note container</div>
  </Note>
);
test('Note renders correctly', async () => {

  const component = renderer.create(note);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const wrapper = shallow(note);
  expect(wrapper).toMatchSnapshot();

  wrapper.setProps({ play: true, });
  expect(wrapper).toMatchSnapshot();
});

test('Note handles onMouseDown and onMouseUp correctly', async () => {
  // TODO: The component needs to be loaded for mousedown and mouseup to be attached
  // const component = renderer.create(note);
  // let tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
  //
  // await tree.props.onMouseDown();
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
  //
  // await tree.props.onMouseUp();
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});

test('Note handles onTouchStart and onTouchEnd correctly', async () => {
  // TODO: The component needs to be loaded for onTouchStart and onTouchEnd to be attached
  // const component = renderer.create(note);
  // let tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
  //
  // await tree.props.onMouseDown();
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
  //
  // await tree.props.onMouseUp();
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});
