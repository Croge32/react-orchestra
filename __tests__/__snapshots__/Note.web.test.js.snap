exports[`test Note handles onMouseDown and onMouseUp correctly 1`] = `
<div
  className=" "
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note handles onMouseDown and onMouseUp correctly 2`] = `
<div
  className=" ro-note-playing"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note handles onMouseDown and onMouseUp correctly 3`] = `
<div
  className=" ro-note-playing"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note handles onTouchStart and onTouchEnd correctly 1`] = `
<div
  className=" "
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note handles onTouchStart and onTouchEnd correctly 2`] = `
<div
  className=" ro-note-playing"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note handles onTouchStart and onTouchEnd correctly 3`] = `
<div
  className=" ro-note-playing"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note renders correctly 1`] = `
<div
  className=" "
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}>
  <div>
    Note container
  </div>
</div>
`;

exports[`test Note renders correctly 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className=" "
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}>
    <div>
        Note container
    </div>
</div>,
  "nodes": Array [
    <div
      className=" "
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}>
      <div>
            Note container
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Note
        instrumentName="acoustic_grand_piano"
        name="A3"
        onNoteLoaded={[Function]}
        play={false}>
        <div>
                Note container
        </div>
</Note>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Note {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onClickStart": [Function],
        "playingBuffers": Array [],
        "props": Object {
          "children": <div>
            Note container
</div>,
          "instrumentName": "acoustic_grand_piano",
          "name": "A3",
          "onNoteLoaded": [Function],
          "play": false,
        },
        "refs": Object {},
        "startPlayingNote": [Function],
        "state": Object {
          "isLoaded": true,
          "isPlaying": false,
        },
        "stopPlayingNote": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className=" "
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}>
          <div>
                    Note container
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className=" "
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}>
          <div>
                    Note container
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Note
    instrumentName="acoustic_grand_piano"
    name="A3"
    onNoteLoaded={[Function]}
    play={false}>
    <div>
        Note container
    </div>
</Note>,
}
`;

exports[`test Note renders correctly 3`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className=" ro-note-playing"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}>
    <div>
        Note container
    </div>
</div>,
  "nodes": Array [
    <div
      className=" ro-note-playing"
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}>
      <div>
            Note container
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Note
        instrumentName="acoustic_grand_piano"
        name="A3"
        onNoteLoaded={[Function]}
        play={true}>
        <div>
                Note container
        </div>
</Note>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Note {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onClickStart": [Function],
        "playingBuffers": Array [],
        "props": Object {
          "children": <div>
            Note container
</div>,
          "instrumentName": "acoustic_grand_piano",
          "name": "A3",
          "onNoteLoaded": [Function],
          "play": true,
        },
        "refs": Object {},
        "startPlayingNote": [Function],
        "state": Object {
          "isLoaded": true,
          "isPlaying": true,
        },
        "stopPlayingNote": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className=" ro-note-playing"
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}>
          <div>
                    Note container
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className=" ro-note-playing"
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}>
          <div>
                    Note container
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Note
    instrumentName="acoustic_grand_piano"
    name="A3"
    onNoteLoaded={[Function]}
    play={true}>
    <div>
        Note container
    </div>
</Note>,
}
`;
